pandora-client-web/src/components/settings/securitySettings.tsx

```diff
@@ -0,0 +1,327 @@
+			</Row>
+		</Column>
+	);
+}
+
+function ExtendCurrentSession({ token }: { token: AuthToken; }): ReactElement {
+	const [show, setShow] = React.useState(false);
+
+	const hide = React.useCallback(() => {
+		setShow(false);
+		return true;
+	}, []);
+
+	useKeyDownEvent(hide, 'Escape');
+
+	if (!show)
+		return <Button onClick={ () => setShow(true) } title='Extend current session'>Extend Session</Button>;
+
+	return (
+		<>
+			<Button className='slim' onClick={ hide } title='Extend current session'>Extend Session</Button>
+			<ExtendCurrentSessionDialog token={ token } hide={ hide } />
+		</>
+	);
+}
+
+function SessionExpire({ token }: { token: AuthToken; }): ReactElement {
+	const now = useCurrentTime(60_000);
+
+	return (
+		<p>Your session will expire in { FormatTimeInterval(token.expires - now, 'two-most-significant') }</p>
+	);
+}
+
+function ExtendCurrentSessionDialog({ token, hide }: { token: AuthToken; hide: () => boolean; }): ReactElement {
+	const directory = useDirectoryConnector();
+	const [password, setPassword] = React.useState('');
+
+	const [extend, processing] = useAsyncEvent(
+		async () => {
+			const passwordSha512 = await PrehashPassword(password);
+			return await directory.awaitResponse('extendLoginToken', { passwordSha512 });
+		},
+		({ result }) => {
+			if (result === 'ok') {
+				toast('Session extended', TOAST_OPTIONS_SUCCESS);
+				hide();
+			} else {
+				toast('Invalid password', TOAST_OPTIONS_ERROR);
+				setPassword('');
+			}
+		},
+	);
+	const onSubmit = React.useCallback((ev: React.FormEvent) => {
+		ev.preventDefault();
+		extend();
+	}, [extend]);
+
+	return (
+		<ModalDialog>
+			<Form dirty={ false } onSubmit={ onSubmit }>
+				<SessionExpire token={ token } />
+				<FormField>
+					<label htmlFor='extend-current-session-password'>Password</label>
+					<input
+						type='password'
+						id='extend-current-session-password'
+						autoComplete='current-password'
+						value={ password }
+						onChange={ (e) => setPassword(e.target.value) }
+					/>
+				</FormField>
+				<br />
+				<Row>
+					<Button onClick={ hide } disabled={ processing }>Cancel</Button>
+					<Button type='submit' disabled={ processing }>Extend</Button>
+				</Row>
+			</Form>
+		</ModalDialog>
+	);
+}
+
+interface PasswordChangeFormData {
+	oldPassword: string;
+	newPassword: string;
+	newPasswordConfirm: string;
+}
+
+function PasswordChange({ account }: { account: IDirectoryAccountInfo; }): ReactElement {
+	const directoryConnector = useDirectoryConnector();
+	const [invalidPassword, setInvalidPassword] = React.useState('');
+
+	const {
+		formState: { errors, submitCount, isSubmitting },
+		reset,
+		getValues,
+		handleSubmit,
+		register,
+		trigger,
+	} = useForm<PasswordChangeFormData>({ shouldUseNativeValidation: true, progressive: true });
+
+	React.useEffect(() => {
+		if (invalidPassword) {
+			void trigger();
+		}
+	}, [invalidPassword, trigger]);
```

Comment:
This only re-runs the validation - the `invalidPassword´ variable is actually never displayed anywhere, so there is no feedback at all if user entered invalid current password.